### ISSUE TITLE
router-sink Application Will Not Route JSON Array Payloads with Spring Boot 2.5.5

### ISSUE THREAD

**Issue by fiidim** [spring-cloud/stream-applications#198]
I've been hunting this issue down for a week now.   If I clone router-sink and and build with `Spring Boot 2.5.4` and dependencies, there is no issue.  When using router-sink built with `2.5.5` and dependencies, the messages do not get routed.

If a `processor` has a signature of `Function<Object, Object[]>`, the `Object[]` portion is serialized to a `byte[]` array.  If this payload is sent to `router-sink`, the message is eaten and not routed anywhere.  Debug logging shows no errors, and tails off somewhere after deserialization.  I assume the issue is either related to dependencies of `2.5.5`, or that it's JSON content.

I wrote a `test` for `router-sink` application that proves my theory (add @Test into `RouterSinkIntegrationTests`).
This test shows a single element payload (success), followed by an array payload (failed), all sent as `byte[]`.
As a side note, I believe that the `router-sink` tests should probably be written as `spring.cloud.function.definition=byteArrayTextToString|routerSinkConsumer` (note the `byteArrayTextToString`), since this is how the sequence will run when deployed.  

```java
@Test
public void testByteMessageDefaultRouter() {
	byte[] jsonArrayBytes = "[{\"id\":\"foo\"},{\"id\":\"bar\"}]".getBytes();
	byte[] jsonSingleBytes = "{\"id\":\"foo\"}".getBytes();

	try (ConfigurableApplicationContext context = new SpringApplicationBuilder(
			TestChannelBinderConfiguration.getCompleteConfiguration(RouterSinkTestApplication.class))
			.web(WebApplicationType.NONE)
			.run("--spring.cloud.function.definition=byteArrayTextToString|routerSinkConsumer",
					"--router.expression=headers['X-EndPoint'].toLowerCase()",
					"--logging.level.org.springframework=DEBUG")) {

		InputDestination processorInput = context.getBean(InputDestination.class);

		Message<?> messageSingle = MessageBuilder.withPayload(jsonSingleBytes)
				.setHeader("X-EndPoint", "CONFIGURATION")
				.setHeader("contentType", "application/json")
				.build();
		processorInput.send(messageSingle);

		OutputDestination processorOutput = context.getBean(OutputDestination.class);
		Message<byte[]> sourceMessage = processorOutput.receive(10000, "configuration");
		assertThat(sourceMessage).isNotNull();
		assertThat(sourceMessage.getPayload()).isEqualTo(jsonSingleBytes);

		Message<?> messageArray = MessageBuilder.withPayload(jsonArrayBytes)
				.setHeader("X-EndPoint", "CONFIGURATION")
				.setHeader("contentType", "application/json")
				.build();
		processorInput.send(messageArray);

		sourceMessage = processorOutput.receive(10000, "configuration");
		assertThat(sourceMessage).isNotNull();
		assertThat(sourceMessage.getPayload()).isEqualTo(jsonArrayBytes);
	}
}
```

**Comment by fiidim** [comment]
Looking deeper this appears to be caused by a recent change in `spring-cloud-function` to `org.springframework.cloud.function.context.catalog.SimpleFunctionRegistry`, and the `fluxifyInputIfNecessary` function.   There is a new code block that causes array payloads to be returned as a list of messages, rather than a message of lists.

Seems like the exclusion criteria might be lacking.

```java
if (!this.isRoutingFunction() && !(input instanceof Publisher)) {
    Object payload = input;
    if (input instanceof Message) {
        payload = ((Message) input).getPayload();
    }
    if (JsonMapper.isJsonStringRepresentsCollection(payload) && !FunctionTypeUtils.isTypeCollection(this.inputType)) {
        payload = jsonMapper.fromJson(payload, List.class);
        MessageHeaders headers = ((Message) input).getHeaders();
        input = ((List) payload).stream()
                .map(p -> MessageBuilder.withPayload(p).copyHeaders(headers).build())
                .collect(Collectors.toList());
    }
}
```

**Comment by fiidim** [comment]
I also opened up a bug with `spring-cloud-function` [here](https://github.com/spring-cloud/spring-cloud-function/issues/762)

**Comment by onobc** [comment]
@fiidim this has been fixed in the latest spring-cloud-function. We will update the sample apps to this version as part of https://github.com/spring-cloud/stream-applications/issues/268 . I will ping you again in this ticket once that work has been done and snapshots are available. At that time we would ask you to verify if it has been fixed. 

Thanks

**Comment by onobc** [comment]
Closing as this has been updated and will be released in `2021.1.2`.
